### DISCUSSION TITLE
subscribe using wildcards

### DISCUSSION THREAD

**Discussion by tnickels** [uNetworking/uWebSockets.js#893] [General]
I originally wrote my websocket server using uWebSockets 19.5.0. I previously went to upgrade to v20+ but it broke. I have just gone back to upgrade to the latest version and I have found the problem, but I do not understand why. I'm hoping someone can answer why my wildcard subscription does not work in versions 20 and above.

I have set up the web server to accept dynamically added topics to a specifically registered room. During the open hook I subscribe the incoming web socket to all topics from that room using #   as per this line ws.subscribe(ws.wsRoomToken+'#');

```js
// Change to SSLApp on deployment
const app = uWS.App({
    // key_file_name: '../ssl.key',
    // cert_file_name: '../ssl.cert'
})
    .ws('/'+wsServerID, {
        compression: uWS.SHARED_COMPRESSOR,
        idleTimeout: 32,
        upgrade: (res, req, context) => {
            res.upgrade({
                    room: req.getQuery('r'),
                },
                req.getHeader('sec-websocket-key'),
                req.getHeader('sec-websocket-protocol'),
                req.getHeader('sec-websocket-extensions'),
                context);
        },
        open: (ws) => {
            //called when a client opens a ws connection with the server
            ws.id = uuid.v4();
            console.log('made new client connection in room: '+ws.room);
            ws.wsRoomToken = ws.room+'/';
            
            ws.subscribe(MESSAGE_ENUM.BROADCAST_MESSAGE);

            ws.subscribe(ws.wsRoomToken+'#');

            let pubMsg = {
                type: ws.wsRoomToken+MESSAGE_ENUM.CLIENT_CONNECTED,
                body: {
                    id: ws.id,
                    name: ws.username
                }
            }

            // send to *all* subscribed sockets
            app.publish(ws.wsRoomToken+MESSAGE_ENUM.CLIENT_CONNECTED, JSON.stringify(pubMsg));

       }
      ....
})
```

when I back date to v19.5.0 it all works. If I remove the # and enter a known subscription topic it works in versions 20 and above. I thought the wildcard character or function had changed but this block of info in the uws.js file remains the same

```
    /** Subscribe to a topic in MQTT syntax.
     *
     * MQTT syntax includes things like "root/child/+/grandchild" where "+" is a
     * wildcard and "root/#" where "#" is a terminating wildcard.
     *
     * Read more about MQTT.
    */
    subscribe(topic: RecognizedString) : boolean;
```

Can someone please guide me on how to get the wildcard working in versions 20 and above

Thanks

**Comment by e3dio** [comment]
First place to check would be v20 release notes https://github.com/uNetworking/uWebSockets/releases/tag/v20.0.0

Which happens to mention exactly this in the first line. It is very easy to get the same behavior, just change around your subscriptions a bit. For your example you would subscribe to `ws.wsRoomToken+MESSAGE_ENUM.CLIENT_CONNECTED` on open

**Reply by tnickels** [reply]
Thank you. I'm using uWebSocketjs so I wasn't looking at the release notes for uWebSockets
